crates/trident/src/subsystems/extensions/mod.rs

```diff
@@ -0,0 +1,792 @@
+        let mut dirs = HashSet::new();
+        dirs.extend(
+            ctx.spec
+                .os
+                .sysexts
+                .iter()
+                .map(|ext| ext.path.clone().unwrap_or(DEFAULT_SYSEXT_DIRECTORY.into())),
+        );
+        dirs.extend(
+            ctx.spec
+                .os
+                .confexts
+                .iter()
+                .map(|ext| ext.path.clone().unwrap_or(DEFAULT_CONFEXT_DIRECTORY.into())),
+        );
+        for dir_path in dirs {
+            let fs_device_id = ctx
+                .spec
+                .storage
+                .path_to_mount_point_info(&dir_path)
+                .and_then(|mp| mp.device_id)
+                .structured(InvalidInputError::from(
+                    HostConfigurationDynamicValidationError::ExtensionImageNotOnABVolume {
+                        path: dir_path.display().to_string(),
+                    },
+                ))?;
+            if ctx.get_ab_volume_pair(fs_device_id).is_err() {
+                return Err(TridentError::new(InvalidInputError::from(
+                    HostConfigurationDynamicValidationError::ExtensionImageNotOnABVolume {
+                        path: dir_path.display().to_string(),
+                    },
+                )));
+            };
+        }
+        Ok(())
+    }
+
+    fn prepare(&mut self, _ctx: &EngineContext) -> Result<(), TridentError> {
+        Ok(())
+    }
+
+    fn provision(&mut self, ctx: &EngineContext, mount_path: &Path) -> Result<(), TridentError> {
+        // Define staging directory, in which extension images will be downloaded.
+        let staging_dir = path::join_relative(mount_path, EXTENSION_IMAGE_STAGING_DIRECTORY);
+
+        // Download new extension images. Mount and process all extension images.
+        self.populate_extensions(ctx, &staging_dir)
+            .structured(InternalError::PopulateExtensionImages)?;
+
+        // TODO: Determine which images need to be removed and which should be added.
+
+        // TODO: Copy extension images to their proper locations.
+
+        // TODO: Clean-up staging directory.
+
+        Ok(())
+    }
+
+    fn configure(&mut self, _ctx: &EngineContext) -> Result<(), TridentError> {
+        Ok(())
+    }
+}
+
+impl ExtensionsSubsystem {
+    fn populate_extensions(
+        self: &mut ExtensionsSubsystem,
+        ctx: &EngineContext,
+        staging_dir: &Path,
+    ) -> Result<(), Error> {
+        let timeout = Duration::from_secs(
+            ctx.spec
+                .internal_params
+                .get_u64(COSI_HTTP_CONNECTION_TIMEOUT_SECONDS)
+                .and_then(|timeout| timeout.ok())
+                .unwrap_or(10),
+        );
+
+        // Create temporary directory in which to download extension images
+        // before copying them to their final path.
+        if !staging_dir.exists() {
+            fs::create_dir_all(staging_dir).with_context(|| {
```

Comment:
should the format message use staging_dir instead of EXTENSION_IMAGE_STAGING_DIRECTORY?